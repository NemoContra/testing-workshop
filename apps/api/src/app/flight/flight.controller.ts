import { Controller, Get, Query } from '@nestjs/common';
import { FlightService } from './flight.service';
import {
  FIND_FLIGHT_FROM,
  FIND_FLIGHT_TO,
  Flight,
} from '@testing-workshop/shared/util/api-interfaces';

@Controller('flight')
export class FlightController {
  constructor(private flightService: FlightService) {}

  @Get()
  getFlights(
    @Query(FIND_FLIGHT_FROM) from?: string,
    @Query(FIND_FLIGHT_TO) to?: string
  ): Flight[] {
    return this.flightService.getFlights({ from, to });
  }
}
