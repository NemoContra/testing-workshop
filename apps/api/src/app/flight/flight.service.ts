import { Injectable } from '@nestjs/common';
import {
  SearchFlightsQueryParams,
  Flight,
} from '@testing-workshop/shared/util/api-interfaces';
import { flights } from './data/flights';

const findOptional = (item: string, optional?: string): boolean =>
  !optional || item.toLowerCase().includes(optional.toLowerCase());

@Injectable()
export class FlightService {
  getFlights({ from, to }: Partial<SearchFlightsQueryParams>): Flight[] {
    return flights
      .filter((flight: Flight) => findOptional(flight.from, from))
      .filter((flight: Flight) => findOptional(flight.to, to));
  }
}
