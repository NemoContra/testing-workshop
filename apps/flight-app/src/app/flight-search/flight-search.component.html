<div class="card">
  <div class="header">
    <h2 class="title">Flight Search</h2>
  </div>
  <div class="content">
    <form [formGroup]="searchFlightsFormGroup">
      <div class="form-group">
        <label>From:</label>
        <input formControlName="from" class="form-control" />
      </div>
      <div class="form-group">
        <label>To:</label>
        <input formControlName="to" name="to" class="form-control" />
      </div>

      <div *ngIf="flights$ | async as flights">
        {{ flights.length }} flights found!

        <div class="flight-cards">
          <flight-card
            *ngFor="let flight of flights"
            [flight]="flight"
          ></flight-card>
        </div>
      </div>
    </form>
  </div>
</div>
