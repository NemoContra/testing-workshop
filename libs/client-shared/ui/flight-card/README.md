# client-shared-ui-flight-card

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test client-shared-ui-flight-card` to execute the unit tests.
