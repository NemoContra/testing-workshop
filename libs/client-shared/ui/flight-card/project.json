{
  "projectType": "library",
  "sourceRoot": "libs/client-shared/ui/flight-card/src",
  "prefix": "flight-card",
  "targets": {
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/client-shared/ui/flight-card"],
      "options": {
        "jestConfig": "libs/client-shared/ui/flight-card/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/client-shared/ui/flight-card/**/*.ts",
          "libs/client-shared/ui/flight-card/**/*.html"
        ]
      }
    }
  },
  "tags": ["scope:client-shared", "type:ui"]
}
