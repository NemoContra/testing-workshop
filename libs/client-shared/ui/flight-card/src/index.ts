export * from './lib/client-shared-ui-flight-card.component';
