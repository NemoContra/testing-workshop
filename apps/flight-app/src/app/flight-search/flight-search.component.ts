import { ChangeDetectionStrategy, Component, OnDestroy } from '@angular/core';
import {
  debounceTime,
  distinctUntilChanged,
  Observable,
  Subject,
  switchMap,
  takeUntil,
} from 'rxjs';
import {
  Flight,
  SearchFlightsQueryParams,
} from '@testing-workshop/shared/util/api-interfaces';
import { FlightService } from '../services/flight.service';
import { FormControl, FormGroup } from '@angular/forms';

@Component({
  selector: 'flight-flight-search',
  templateUrl: './flight-search.component.html',
  styleUrls: ['./flight-search.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FlightSearchComponent implements OnDestroy {
  private destroy$$ = new Subject<void>();

  searchFlightsFormGroup = new FormGroup({
    from: new FormControl(''),
    to: new FormControl(''),
  });

  flights$: Observable<Flight[]> =
    this.searchFlightsFormGroup.valueChanges.pipe(
      debounceTime(250),
      distinctUntilChanged(),
      switchMap((searchFlightsQueryParams) =>
        this.searchFlights(searchFlightsQueryParams as SearchFlightsQueryParams)
      ),
      takeUntil(this.destroy$$)
    );

  constructor(private flightService: FlightService) {}

  ngOnDestroy(): void {
    this.destroy$$.next();
  }

  private searchFlights(
    searchFlightsQueryParams: SearchFlightsQueryParams
  ): Observable<Flight[]> {
    return this.flightService.searchFlights(searchFlightsQueryParams);
  }
}
