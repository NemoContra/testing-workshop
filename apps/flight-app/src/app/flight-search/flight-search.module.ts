import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FlightSearchComponent } from './flight-search.component';
import { ReactiveFormsModule } from '@angular/forms';
import { ClientSharedUiFlightCardComponent } from '@testing-workshop/client-shared/ui/flight-card';

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    ClientSharedUiFlightCardComponent,
  ],
  declarations: [FlightSearchComponent],
  exports: [FlightSearchComponent],
})
export class FlightSearchModule {}
