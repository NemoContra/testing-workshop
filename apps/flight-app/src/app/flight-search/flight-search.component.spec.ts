import { createComponentFactory, mockProvider } from '@ngneat/spectator/jest';
import { FlightSearchComponent } from './flight-search.component';
import { MockComponent } from 'ng-mocks';
import { ClientSharedUiFlightCardComponent } from '@testing-workshop/client-shared/ui/flight-card';
import { FlightService } from '../services/flight.service';
import { ReactiveFormsModule } from '@angular/forms';
import { SearchFlightsQueryParams } from '@testing-workshop/shared/util/api-interfaces';
import { fakeAsync, tick } from '@angular/core/testing';
import { testIdSelector } from '@testing-workshop/shared/util/test-helpers';
import { EMPTY, of } from 'rxjs';
import { flights } from '@testing-workshop/shared/util/mock-data';

describe('FlightSearchComponent', () => {
  const createComponent = createComponentFactory({
    component: FlightSearchComponent,
    imports: [
      MockComponent(ClientSharedUiFlightCardComponent),
      ReactiveFormsModule,
    ],
    providers: [
      mockProvider(FlightService, {
        searchFlights: jest.fn().mockReturnValue(EMPTY),
      }),
    ],
  });

  afterEach(() => jest.clearAllMocks());

  const selectors = {
    fromInput: testIdSelector('from'),
    toInput: testIdSelector('to'),
  } as const;

  it('should match snapshot without flights', () => {
    const spectator = createComponent();
    expect(spectator.fixture).toMatchSnapshot();
  });

  it.each`
    from         | to
    ${'Graz'}    | ${undefined}
    ${'Graz'}    | ${'Hamburg'}
    ${undefined} | ${'Hamburg'}
  `(
    'should search a flight from $from to $to',
    fakeAsync(({ from, to }: SearchFlightsQueryParams) => {
      const spectator = createComponent();
      const flightService = spectator.inject(FlightService);

      from && spectator.typeInElement(from, selectors.fromInput);
      to && spectator.typeInElement(to, selectors.toInput);

      // Debounce waits for 250ms so at first nothing happens
      expect(flightService.searchFlights).not.toHaveBeenCalled();

      tick(250);

      // After 250ms the call happens
      expect(flightService.searchFlights).toMatchSnapshot();
    })
  );

  it('should render a list of flights and show the amount found', fakeAsync(() => {
    const spectator = createComponent();
    const flightService = spectator.inject(FlightService);

    flightService.searchFlights.mockReturnValue(of(flights));
    spectator.typeInElement('Graz', selectors.fromInput);
    tick(250);
    spectator.detectChanges();

    expect(spectator.query(testIdSelector('content'))).toMatchSnapshot();
  }));
});
