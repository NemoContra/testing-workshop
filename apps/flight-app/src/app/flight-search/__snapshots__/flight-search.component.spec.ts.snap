// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightSearchComponent should match snapshot without flights 1`] = `
<flight-flight-search
  __ngContext__="0"
  destroy$$={[Function Subject]}
  flightService={[Function Object]}
  flights$={[Function AnonymousSubject]}
  searchFlightsFormGroup={[Function FormGroup]}
>
  <div
    class="card"
  >
    <div
      class="header"
    >
      <h2
        class="title"
      >
        Search Flights
      </h2>
    </div>
    <div
      class="content"
      data-tid="content"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="form-group"
        >
          <label>
            From:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            data-tid="from"
            formcontrolname="from"
          />
        </div>
        <div
          class="form-group"
        >
          <label>
            To:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            data-tid="to"
            formcontrolname="to"
          />
        </div>
      </form>
    </div>
  </div>
</flight-flight-search>
`;

exports[`FlightSearchComponent should render a list of flights and show the amount found 1`] = `
<div
  class="content"
  data-tid="content"
>
  <form
    class="ng-untouched ng-valid ng-dirty"
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label>
        From:
      </label>
      <input
        class="form-control ng-untouched ng-valid ng-dirty"
        data-tid="from"
        formcontrolname="from"
      />
    </div>
    <div
      class="form-group"
    >
      <label>
        To:
      </label>
      <input
        class="form-control ng-untouched ng-pristine ng-valid"
        data-tid="to"
        formcontrolname="to"
      />
    </div>
    <div>
       2 flights found! 
      <div
        class="flight-cards"
      >
        <flight-card />
        <flight-card />
        
      </div>
    </div>
    
  </form>
</div>
`;

exports[`FlightSearchComponent should search a flight from Graz to Hamburg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "from": "Graz",
        "to": "Hamburg",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Observable {
        "_subscribe": [Function],
      },
    },
  ],
}
`;

exports[`FlightSearchComponent should search a flight from Graz to undefined 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "from": "Graz",
        "to": "",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Observable {
        "_subscribe": [Function],
      },
    },
  ],
}
`;

exports[`FlightSearchComponent should search a flight from undefined to Hamburg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "from": "",
        "to": "Hamburg",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Observable {
        "_subscribe": [Function],
      },
    },
  ],
}
`;
